presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,541 @@
+                int probe = nextProbe(probeCount);
+                bucketId = nextBucketId(originalBucket, newMask, probe);
+                probeCount++;
+            }
+
+            // record the mapping
+            newBuckets.set(bucketId, i);
+        }
+
+        buckets = newBuckets;
+        bucketCount = newBucketCount;
+        maxFill = calculateMaxFill(newBucketCount, MAX_FILL_RATIO);
+        mask = newMask;
+        resizeNodeArrays(newBucketCount);
+    }
+
+    private int nextProbe(int probeCount)
+    {
+        switch (PROBE_TYPE) {
+            case LINEAR:
+                return nextProbeLinear(probeCount);
+            case SUM_OF_COUNT:
+                return nextSumOfCount(probeCount);
+            case SUM_OF_SQUARE:
+                return nextSumOfSquares(probeCount);
+        }
+
+        throw new IllegalArgumentException(String.format("invalid enum %s", PROBE_TYPE));
+    }
+
+    //parallel arrays with data for
+    private void resizeNodeArrays(int newBucketCount)
+    {
+        // every per-bucket array needs to be updated
+        counts.ensureCapacity(newBucketCount);
+        valuePositions.ensureCapacity(newBucketCount);
+        nextPointers.ensureCapacity(newBucketCount);
+        valueHashList.ensureCapacity(newBucketCount);
+        groupIds.ensureCapacity(newBucketCount);
+    }
+
+    private long getHash(long groupId, long valueRawHash)
+    {
+        return (murmurHash3(valueRawHash) ^ murmurHash3(groupId));
+    }
+
+    private int getBucketIdForNode(int nodePointer, int mask)
+    {
+        long valueRawHash = valueHashList.get(nodePointer); // without mask
+        int bucketId = (int) (valueRawHash & mask);
+
+        return bucketId;
+    }
+
+    //short-lived abstraction that is basically a position into parallel arrays that we can treat as one data structure
+    private class ValueNode
+    {
+        // index into parallel arrays that are fields of a "node" in our hash table (eg counts, valuePositions)
+        private final int nodePointer;
+
+        /**
+         * @param nodePointer - index/pointer into parallel arrays of data structs
+         * @param isEmpty
+         */
+        ValueNode(int nodePointer)
+        {
+            checkState(nodePointer > -1, "ValueNode must point to a non-empty node");
+            this.nodePointer = nodePointer;
+        }
+
+        long getCount()
+        {
+            return counts.get(nodePointer);
+        }
+
+        int getValuePosition()
+        {
+            return valuePositions.get(nodePointer);
+        }
+
+        void add(long count)
+        {
+            counts.add(nodePointer, count);
+        }
+
+        /**
+         * given an output outputBlockBuilder, writes one row (key -> count) of our histogram
+         *
+         * @param valuesBlock - values.build() is called externally
+         */
+        void writeNodeAsRow(Block valuesBlock, BlockBuilder outputBlockBuilder)
```

Comment:
Technically, this is not a proper "row". A row needs to use RowBlock.

Reply:
rename? writeNodeAsBlock?

Reply:
renamed